### ISSUE TITLE
Moq1500 never analyzes constructor bodies (xUnit constructor-setup pattern missed); dead PropertyReference branch; double Descendants() walk per declarator

### ISSUE THREAD

**Issue by rjmurillo** [rjmurillo/moq.analyzers#1253]
## Summary

An internal code audit found two defects in `MockRepositoryVerifyAnalyzer` (Moq1500):

1. **Medium FN.** `GetContainingMember` accepts only `OperationKind.MethodBody` (and a dead `OperationKind.PropertyReference`), but a variable declarator inside a **constructor** roots at `OperationKind.ConstructorBody`. The walk returns `null` and the analyzer bails — so the very common xUnit pattern of creating a `MockRepository` in the test-class constructor is never analyzed.
2. **Perf (audit L3).** `HasCreateCallsForRepository` and `HasVerifyCallsForRepository` each perform a full `Descendants()` walk of the containing member per repository declarator; a single pass can compute both flags.

Root-kind facts were verified empirically against the pinned Roslyn (4.8.0) with declarators in each container:

| Container of the local declarator | Root operation kind (`Parent == null`) |
|---|---|
| Method body | `MethodBody` |
| Constructor body | `ConstructorBody` |
| Property getter/setter accessor | `MethodBody` |
| Local function | `MethodBody` (of the outer method; chain passes through `LocalFunction`) |
| Field initializer | not applicable (no `IVariableDeclaratorOperation` is produced for field declarators, so the callback never fires — the documented field-level limitation) |

Two consequences: (a) property accessors already work today (they root at `MethodBody`), so the fix needs no special accessor handling — but the behavior should be pinned by a test; (b) `OperationKind.PropertyReference` can never appear on the ancestor path of a variable declarator (a property reference is an expression and can only be a descendant), so that branch is provably dead.

Related but separate: #986 tracks `VerifyAll()` not suppressing Moq1500 and the cross-method scope limitation. This issue must **not** change either behavior (they are pinned by `TODO(#986)` tests at `tests/Moq.Analyzers.Test/MockRepositoryVerifyAnalyzerTests.cs:227` and `:245`).

## Current behavior

`src/Analyzers/MockRepositoryVerifyAnalyzer.cs:97-114`:

```csharp
    /// <summary>
    /// Gets the containing method or property for an operation.
    /// </summary>
    private static IOperation? GetContainingMember(IOperation operation)
    {
        IOperation? current = operation;
        while (current != null)
        {
            if (current.Kind == OperationKind.MethodBody || current.Kind == OperationKind.PropertyReference)
            {
                return current;
            }

            current = current.Parent;
        }

        return null;
    }
```

Consumed at `src/Analyzers/MockRepositoryVerifyAnalyzer.cs:77-86`:

```csharp
        // Get the containing method or property
        IOperation? containingMember = GetContainingMember(declarator);
        if (containingMember is null)
        {
            return;
        }

        // Check if there are Create() calls and Verify() calls for this repository in the same scope
        bool hasCreateCalls = HasCreateCallsForRepository(declarator.Symbol, containingMember, knownSymbols);
        bool hasVerifyCalls = HasVerifyCallsForRepository(declarator.Symbol, containingMember, knownSymbols);
```

The analyzer registers for `OperationKind.VariableDeclarator` (`src/Analyzers/MockRepositoryVerifyAnalyzer.cs:56-58`), so callbacks fire for locals in every operation block — including constructor bodies — but the ancestor walk then rejects them.

**Concrete failure input → wrong outcome:**

```csharp
public class MyTests
{
    private readonly object _mock;

    public MyTests()   // xUnit per-test setup
    {
        var repo = new MockRepository(MockBehavior.Strict);   // expected Moq1500 — NOT reported
        _mock = repo.Create<IFoo>();
    }
}
```

The declarator's ancestor chain is `VariableDeclaration <- VariableDeclarationGroup <- Block <- ConstructorBody` (verified); no `MethodBody` is found, `GetContainingMember` returns `null`, and the analyzer silently skips. No Moq1500 is ever produced for constructor bodies.

The two scan helpers at `src/Analyzers/MockRepositoryVerifyAnalyzer.cs:119-132` and `:137-150` each iterate `memberOperation.Descendants()` in full:

```csharp
    private static bool HasCreateCallsForRepository(ILocalSymbol repositorySymbol, IOperation memberOperation, MoqKnownSymbols knownSymbols)
    {
        foreach (IOperation operation in memberOperation.Descendants())
        {
            if (operation is IInvocationOperation invocation &&
                IsValidMockRepositoryCreateCall(invocation, knownSymbols) &&
                GetRepositorySymbolFromCreateCall(invocation)?.Equals(repositorySymbol, SymbolEqualityComparer.Default) == true)
            {
                return true;
            }
        }

        return false;
    }
```

(and the structurally identical `HasVerifyCallsForRepository`). Two full walks per `MockRepository` declarator, per member body.

## Desired behavior

- A `MockRepository` local created in a constructor (or any other member body) is analyzed with the same "Create without Verify in the same member body" rule as method bodies. Constructor positive/negative cases behave exactly like the existing method-body rows.
- The member scope is determined robustly by walking to the **root** of the operation tree (`Parent == null`) instead of matching an allow-list of kinds. This fixes constructors, keeps methods/accessors/local functions identical to today (their root is the same `MethodBody` node the current code finds), removes the dead `PropertyReference` branch, and cannot return null — no silent bail-out path.
- Create/Verify detection for a declarator is computed in a single `Descendants()` pass returning both flags.
- Unchanged (per #986): `VerifyAll()` does not suppress the diagnostic; Verify in a different member than Create does not suppress it.

## Implementation plan

All changes in `src/Analyzers/MockRepositoryVerifyAnalyzer.cs`. `System.ValueTuple` is available on `netstandard2.0`; no new usings needed (`Microsoft.CodeAnalysis.Operations` is already imported at line 1).

1. Replace `AnalyzeVariableDeclaration` (currently lines 61-95) with:

```csharp
    /// <summary>
    /// Analyzes a variable declaration to find MockRepository instances that may need verification.
    /// </summary>
    private static void AnalyzeVariableDeclaration(OperationAnalysisContext context, MoqKnownSymbols knownSymbols)
    {
        if (context.Operation is not IVariableDeclaratorOperation declarator)
        {
            return;
        }

        // Check if this declares a MockRepository variable
        if (!IsValidMockRepositoryDeclaration(declarator, knownSymbols))
        {
            return;
        }

        // Scope the analysis to the enclosing member body: the root of the operation tree
        // (method body, constructor body, accessor body, ...).
        IOperation containingMember = GetRootOperation(declarator);

        // Check if there are Create() calls and Verify() calls for this repository in the same scope
        (bool hasCreateCalls, bool hasVerifyCalls) = GetCreateAndVerifyCallsForRepository(declarator.Symbol, containingMember, knownSymbols);

        // Report diagnostic if Create() calls exist but no Verify() calls
        if (hasCreateCalls && !hasVerifyCalls)
        {
            Location diagnosticLocation = GetDiagnosticLocationForVariableDeclarator(declarator);
            Diagnostic diagnostic = diagnosticLocation.CreateDiagnostic(Rule, declarator.Symbol.Name);
            context.ReportDiagnostic(diagnostic);
        }
    }
```

2. Replace `GetContainingMember` (currently lines 97-114) with:

```csharp
    /// <summary>
    /// Walks to the root of the operation tree containing <paramref name="operation"/>.
    /// For operations inside member bodies this is the member body operation itself
    /// (<see cref="OperationKind.MethodBody"/> for methods, accessors, and local functions;
    /// <see cref="OperationKind.ConstructorBody"/> for constructors).
    /// </summary>
    private static IOperation GetRootOperation(IOperation operation)
    {
        IOperation current = operation;
        while (current.Parent is not null)
        {
            current = current.Parent;
        }

        return current;
    }
```

3. Replace **both** `HasCreateCallsForRepository` (currently lines 116-132) and `HasVerifyCallsForRepository` (currently lines 134-150) with a single method (delete both originals):

```csharp
    /// <summary>
    /// Scans the member body once and reports whether it contains Create() calls and
    /// Verify() calls on the specified repository local.
    /// </summary>
    private static (bool HasCreateCalls, bool HasVerifyCalls) GetCreateAndVerifyCallsForRepository(
        ILocalSymbol repositorySymbol,
        IOperation memberOperation,
        MoqKnownSymbols knownSymbols)
    {
        bool hasCreateCalls = false;
        bool hasVerifyCalls = false;

        foreach (IOperation operation in memberOperation.Descendants())
        {
            if (operation is not IInvocationOperation invocation)
            {
                continue;
            }

            if (!hasCreateCalls
                && IsValidMockRepositoryCreateCall(invocation, knownSymbols)
                && GetRepositorySymbolFromCreateCall(invocation)?.Equals(repositorySymbol, SymbolEqualityComparer.Default) == true)
            {
                hasCreateCalls = true;
            }
            else if (!hasVerifyCalls
                && IsValidMockRepositoryVerifyCall(invocation, knownSymbols)
                && GetRepositorySymbolFromVerifyCall(invocation)?.Equals(repositorySymbol, SymbolEqualityComparer.Default) == true)
            {
                hasVerifyCalls = true;
            }

            if (hasCreateCalls && hasVerifyCalls)
            {
                break;
            }
        }

        return (hasCreateCalls, hasVerifyCalls);
    }
```

   (The `else` is safe: a single invocation cannot be both a `Create` and a `Verify` call, and once a flag is set the corresponding branch is skipped.)

4. Everything else in the file (`IsValidMockRepositoryDeclaration`, `IsValidMockRepositoryVerifyCall`, `IsValidMockRepositoryCreateCall`, `GetRepositorySymbolFromVerifyCall`, `GetRepositorySymbolFromCreateCall`, `GetDiagnosticLocationForVariableDeclarator`, registration) is unchanged.

5. Update `docs/rules/Moq1500.md`: add a constructor example to "Examples of patterns that are flagged by this analyzer" (repository created and used in a test-class constructor without `Verify()` in that constructor). Do not touch `AnalyzerReleases.Shipped.md`.

## Tests

All in `tests/Moq.Analyzers.Test/MockRepositoryVerifyAnalyzerTests.cs`. Reuse the existing `VerifyWithNamespacePrefix` helper (line 356) for the class-shaped sources.

### 1. New MemberData — constructor bodies

```csharp
    public static IEnumerable<object[]> ConstructorBodyTestData()
    {
        return new object[][]
        {
            // Constructor creates and uses the repository without Verify() - diagnostic.
            // This is the common xUnit constructor-setup pattern.
            [
                """
                internal interface IFoo
                {
                    void DoSomething();
                }

                internal class UnitTest
                {
                    private readonly object _mock;

                    public UnitTest()
                    {
                        var {|Moq1500:repository|} = new MockRepository(MockBehavior.Strict);
                        _mock = repository.Create<IFoo>();
                    }
                }
                """,
            ],

            // Constructor creates and verifies - no diagnostic.
            [
                """
                internal interface IFoo
                {
                    void DoSomething();
                }

                internal class UnitTest
                {
                    private readonly object _mock;

                    public UnitTest()
                    {
                        var repository = new MockRepository(MockBehavior.Strict);
                        _mock = repository.Create<IFoo>();
                        repository.Verify();
                    }
                }
                """,
            ],

            // Constructor with Create() but no repository usage after - still a diagnostic
            // (same semantics as the method-body rows).
            [
                """
                internal interface IFoo
                {
                    void DoSomething();
                }

                internal class UnitTest
                {
                    public UnitTest()
                    {
                        var {|Moq1500:repository|} = new MockRepository(MockBehavior.Loose);
                        var fooMock = repository.Create<IFoo>();
                    }
                }
                """,
            ],

            // TODO(#986): VerifyAll() does not suppress the diagnostic (existing pinned
            // behavior, now also pinned for constructor bodies - do NOT change in this issue).
            [
                """
                internal interface IFoo
                {
                    void DoSomething();
                }

                internal class UnitTest
                {
                    public UnitTest()
                    {
                        var {|Moq1500:repository|} = new MockRepository(MockBehavior.Strict);
                        var fooMock = repository.Create<IFoo>();
                        repository.VerifyAll();
                    }
                }
                """,
            ],

            // TODO(#986): analysis is member-scoped; Verify() in another member does not
            // suppress the diagnostic in the constructor (existing pinned limitation).
            [
                """
                internal interface IFoo
                {
                    void DoSomething();
                }

                internal class UnitTest
                {
                    public UnitTest()
                    {
                        var {|Moq1500:repository|} = new MockRepository(MockBehavior.Strict);
                        var fooMock = repository.Create<IFoo>();
                    }

                    private void Cleanup()
                    {
                        var repository = new MockRepository(MockBehavior.Strict);
                        repository.Verify();
                    }
                }
                """,
            ],
        }.WithNamespaces().WithMoqReferenceAssemblyGroups();
    }

    [Theory]
    [MemberData(nameof(ConstructorBodyTestData))]
    public async Task ShouldAnalyzeConstructorBodies(string referenceAssemblyGroup, string @namespace, string testCode)
    {
        await VerifyWithNamespacePrefix(referenceAssemblyGroup, @namespace, testCode);
    }
```

Pre-fix, the three `{|Moq1500:...|}` constructor rows fail (no diagnostic reported) — confirm red before fixing.

### 2. New MemberData — property accessor bodies (pins already-working behavior)

Accessor bodies root at `OperationKind.MethodBody` (verified), so these pass both before and after the change; they pin the behavior so a future refactor cannot regress it silently.

```csharp
    public static IEnumerable<object[]> AccessorBodyTestData()
    {
        return new object[][]
        {
            // Getter creates and uses the repository without Verify() - diagnostic.
            [
                """
                internal interface IFoo
                {
                    void DoSomething();
                }

                internal class UnitTest
                {
                    private object Mocks
                    {
                        get
                        {
                            var {|Moq1500:repository|} = new MockRepository(MockBehavior.Strict);
                            return repository.Create<IFoo>();
                        }
                    }
                }
                """,
            ],

            // Getter creates and verifies - no diagnostic.
            [
                """
                internal interface IFoo
                {
                    void DoSomething();
                }

                internal class UnitTest
                {
                    private object Mocks
                    {
                        get
                        {
                            var repository = new MockRepository(MockBehavior.Strict);
                            var mock = repository.Create<IFoo>();
                            repository.Verify();
                            return mock;
                        }
                    }
                }
                """,
            ],
        }.WithNamespaces().WithMoqReferenceAssemblyGroups();
    }

    [Theory]
    [MemberData(nameof(AccessorBodyTestData))]
    public async Task ShouldAnalyzePropertyAccessorBodies(string referenceAssemblyGroup, string @namespace, string testCode)
    {
        await VerifyWithNamespacePrefix(referenceAssemblyGroup, @namespace, testCode);
    }
```

### 3. Existing tests that must keep passing unchanged

All existing theories in `tests/Moq.Analyzers.Test/MockRepositoryVerifyAnalyzerTests.cs`:

- `ShouldAnalyzeMockRepositoryUsage` (`TestData`, 9 rows) and `ShouldAnalyzeMultipleRepositories` (4 rows) — method bodies; root walk finds the same `MethodBody` node as before.
- `ShouldNotReportForFieldLevelRepositories` — field-level repos still produce no callback (no `IVariableDeclaratorOperation` for fields).
- `ShouldNotReportWhenVerifyIsConditional`, `ShouldNotReportWhenMoqIsNotReferenced`.
- `ShouldDocumentVerifyAllBehavior` (`TODO(#986)` pin at line 227) and `ShouldReportWhenCreateAndVerifyInSeparateMethods` (`TODO(#986)` pin at line 245) — **must not change**; this issue does not alter `VerifyAll`/cross-member semantics.

Moq-version note: `MockRepository`, `Create<T>()`, `Verify()`, and `VerifyAll()` exist in both Moq 4.8.2 and 4.18.4, so all rows multiply cleanly across `WithMoqReferenceAssemblyGroups()`.

## Constraints

- Analyzer assemblies target `netstandard2.0`; Roslyn pinned to Microsoft.CodeAnalysis 4.8 (ADR-003); no new package dependencies. (`ValueTuple` returns are fine on netstandard2.0.)
- Symbol-based detection (ADR-001): unchanged — `MockRepository`/`Create`/`Verify` identification stays via `MoqKnownSymbols.MockRepository`, `MockRepositoryCreate`, `MockRepositoryVerify` (`src/Common/WellKnown/MoqKnownSymbols.cs:246,257,268`) resolved through `TypeProvider.GetOrCreateTypeByMetadataName`; no string comparisons are introduced.
- Tests follow repo patterns: `{|Moq1500:...|}` markup (ID + span on the variable identifier, matching `GetDiagnosticLocationForVariableDeclarator`), `[Theory]` + `[MemberData]`, rows multiplied across 2 namespaces and 2 Moq versions (4.8.2/4.18.4).
- Validation: `dotnet format`, `dotnet build` (0 warnings), `dotnet test --settings ./build/targets/tests/test.runsettings`.
- Update `docs/rules/Moq1500.md` (behavior change: constructor bodies now analyzed); never touch `AnalyzerReleases.Shipped.md`.

## Validation commands

```bash
dotnet format
dotnet build            # must complete with 0 warnings
dotnet test --settings ./build/targets/tests/test.runsettings
```

## Acceptance criteria

- [ ] A `MockRepository` local created and used via `Create<T>()` in a constructor without `Verify()` in that constructor reports Moq1500 on the variable identifier (confirmed failing pre-fix).
- [ ] The same constructor pattern with `repository.Verify()` reports nothing.
- [ ] Constructor + `VerifyAll()` still reports (pins #986 behavior; `TODO(#986)` comment on the row); constructor + `Verify()` in a different member still reports (member-scoped, `TODO(#986)`).
- [ ] Property-accessor rows pass (pinned behavior).
- [ ] `GetContainingMember` is replaced by a root-walk (`GetRootOperation`); the dead `OperationKind.PropertyReference` branch is gone; there is no null-scope bail-out.
- [ ] Create/Verify detection is a single `Descendants()` pass returning both flags; `HasCreateCallsForRepository`/`HasVerifyCallsForRepository` are deleted.
- [ ] All existing `MockRepositoryVerifyAnalyzerTests` theories pass unchanged, including both `TODO(#986)` pins.
- [ ] `docs/rules/Moq1500.md` gains a constructor example; `AnalyzerReleases.Shipped.md` untouched.
- [ ] `dotnet format`, `dotnet build` (0 warnings), and the full test suite pass.

## Out of scope

- #986 itself: `VerifyAll()` suppression and cross-member (e.g. constructor-creates / `Dispose`-verifies) tracking — related but separate; this issue must not change those semantics.
- Field-level repository tracking (`_repository = new MockRepository(...)` as a field) — documented limitation, unchanged.
- Repositories passed to helper methods that call `Verify()` — inherent single-member design limit noted by the audit.
